psydac/api/tests/test_api_2d_scalar_openmp.py

```diff
@@ -0,0 +1,390 @@
+# -*- coding: UTF-8 -*-
+
+import os
+import pytest
+
+from mpi4py import MPI
+from sympy import pi, sin, cos, symbols
+
+from sympde.calculus import grad, dot
+from sympde.calculus import laplace
+from sympde.topology import ScalarFunctionSpace
+from sympde.topology import element_of
+from sympde.topology import NormalVector
+from sympde.topology import Square, Domain
+from sympde.topology import Union
+from sympde.expr     import BilinearForm, LinearForm, integral
+from sympde.expr     import Norm
+from sympde.expr     import find, EssentialBC
+
+from psydac.api.discretization import discretize
+from psydac.api.settings import PSYDAC_BACKEND_GPYCCEL
+
+comm = MPI.COMM_WORLD
+PSYDAC_BACKEND_GPYCCEL = PSYDAC_BACKEND_GPYCCEL.copy()
+PSYDAC_BACKEND_GPYCCEL['openmp'] = True
```

Comment:
we can try to do it using `pytest.mark.parametrize`, but for this, we need to add the backends to both `test_api_2d_scalar.py` and `test_api_2d_scalar_mapping.py`

Reply:
Yes, I think it makes sense do it this way. Maybe first we can split the long `test_api_2d_scalar.py` into some smaller files based on the equation to be solved:
- `test_2d_laplace.py`
- `test_2d_poisson.py`
- `test_2d_biharmonic.py`

And similarly for `test_api_2d_scalar_mapping.py`:
- `test_2d_mapping_laplace.py`
- `test_2d_mapping_poisson.py`
- `test_2d_mapping_biharmonic.py`

I think we can do even better by considering the mapped and non-mapped cases in the same file.

Reply:
yes, that's sounds good to me

Reply:
done, I have kept the mapped and non-mapped cases in different files because we still have different ways of doing them, either by reading a file or creating an analytical mapping

Reply:
Thanks for reorganizing the files, this is going to be very useful! But I think that there is still a lot of duplicated code that can be handled through `@pytest.mark.parametrize`:

1. Set `os.environ['OMP_NUM_THREADS'] = "2"` globally, as it is only used when OpenMP is activated
2. Create three different dictionaries in the file:
    - `backend_python`
    - `backend_gfortran`
    - `backend_gfortran_openmp`
3. Change signature to
```python
@pytest.mark.parametrize('backend',  [backend_python, backend_gfortran, backend_gfortran_openmp])
def test_poisson_2d_dir0_1234(backend):
    ...
```

Reply:
if we set `os.environ['OMP_NUM_THREADS'] = "2"` without activating the backend,  it will then generate a code with OpenMP pragmas, but it will not compile with OpenMP

Reply:
But the generated code depends on the `backend['openmp']` flag, right?

Reply:
the generated code doesn't depend on it, it depends on the flag `OMP_NUM_THREADS`, only the compiled code depends on `backend['openmp']`

Reply:
This does not make much sense to me, because the Python code that we generate already depends on the flags contained in the `backend` dictionary (see type annotations). So I think it is more consistent to only add the OpenMP pragmas if the user asks for them with `backend['openmp'] = True`.

As it is now, what happens if the user has `OMP_NUM_THREADS=4` but forgets to set `backend['openmp'] = True`? Will the generated code yield the correct results? I don't think so...

Reply:
yes, you are right, I will add that condition

Reply:
Thanks! There seems to be an issue in the MPI parallel tests, though...

Reply:
yes, I fixed them